packages/docusaurus/src/client/ClientLifecyclesDispatcher.tsx

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, {
+  useEffect,
+  useImperativeHandle,
+  useLayoutEffect,
+  type ReactNode,
+} from 'react';
+import clientModules from '@generated/client-modules';
+import type {ClientModule} from '@docusaurus/types';
+import type {Location} from 'history';
+
+function dispatchLifecycleAction<K extends keyof ClientModule>(
+  lifecycleAction: K,
+  ...args: Parameters<NonNullable<ClientModule[K]>>
+) {
+  clientModules.forEach((clientModule) => {
+    const lifecycleFunction = (clientModule?.default?.[lifecycleAction] ??
+      clientModule[lifecycleAction]) as
+      | ((...a: Parameters<NonNullable<ClientModule[K]>>) => void)
+      | undefined;
+
+    lifecycleFunction?.(...args);
+  });
+}
+
+function ClientLifecyclesDispatcher(
+  {
+    children,
+    location,
+    previousLocation,
+  }: {
+    children: ReactNode;
+    location: Location;
+    previousLocation: Location | null;
+  },
+  ref: React.ForwardedRef<ClientModule>,
+): JSX.Element {
+  useEffect(() => {
+    if (previousLocation !== location) {
+      const {hash} = location;
+      if (!hash) {
+        window.scrollTo(0, 0);
+      } else {
+        const id = decodeURIComponent(hash.substring(1));
+        const element = document.getElementById(id);
+        element?.scrollIntoView();
+      }
+    }
+  });
+  useLayoutEffect(() => {
+    dispatchLifecycleAction('onRouteUpdate', {previousLocation, location});
+  });
```

Comment:
I'm not sure whether putting this in `useLayoutEffect` or `useEffect` is better semantic. Because client modules don't have access to the React context and they contain purely imperative logic (again, I don't think analytics can possibly block rendering anyways), putting them in either places make sense.

Reply:
Note that I **intentionally** left out the `pathname === pathname` check. I think this should be handled on the downstream because who knows? Some client modules may find this useful.